src/mapgen_functions.cpp

```diff
@@ -977,382 +980,319 @@ void mapgen_fungal_flowers(map *m, oter_id, mapgendata dat, int, float)
+
+    // which way should our roads curve, based on neighbor roads?
+    int curvedir_nesw[4] = {};
+    for (int dir = 0; dir < 4; dir++) { // N E S W
+        if (roads_nesw[dir] == false) continue;
+        if (otermap[dat.t_nesw[dir]].id_base != "road") continue;
+
+        // n_* contain details about the neighbor being considered
+        bool n_roads_nesw[4] = {};
+        //TODO figure out how to call this function without creating a new oter_id object
+        int n_num_dirs = terrain_type_with_suffix_to_nesw_array(oter_id(otermap[dat.t_nesw[dir]].id), n_roads_nesw);
+        // if 2-way neighbor has a road facing us
+        if (n_num_dirs==2 && n_roads_nesw[(dir+2)%4]) {
+            // curve towards the direction the neighbor turns
+            if (n_roads_nesw[(dir-1+4)%4]) curvedir_nesw[dir]--; // our road curves counterclockwise
+            if (n_roads_nesw[(dir+1  )%4]) curvedir_nesw[dir]++; // our road curves clockwise
+        }
+    }
+
+    // calculate how far to rotate the map so we can work with just one orientation
+    // also keep track of diagonal roads and plazas
+    int rot = 0;
+    bool diag = false;
+    int plaza_dir = -1;
+    bool fourways_neswx[8] = {};
+    //TODO reduce amount of logical/conditional constructs here
+    switch (num_dirs) {
+        case 4: // 4-way intersection
+            for (int dir = 0; dir < 8; dir++) {
+                fourways_neswx[dir] =
+                    ( otermap[dat.t_nesw[dir]].id == "road_nesw" ||
+                      otermap[dat.t_nesw[dir]].id == "road_nesw_manhole" );
+            }
+            if      (std::equal(std::begin(fourways_neswx), std::end(fourways_neswx), std::begin({1,1,1,1,1,1,1,1})))
+                plaza_dir = 8; // surrounded by 4ways
+            else if (std::equal(std::begin(fourways_neswx), std::end(fourways_neswx), std::begin({0,1,1,0,0,1,0,0})))
+                plaza_dir = 7; // northwest corner of rotary/plaza
+            else if (std::equal(std::begin(fourways_neswx), std::end(fourways_neswx), std::begin({1,1,0,0,1,0,0,0})))
+                plaza_dir = 6;
+            else if (std::equal(std::begin(fourways_neswx), std::end(fourways_neswx), std::begin({1,0,0,1,0,0,0,1})))
+                plaza_dir = 5;
+            else if (std::equal(std::begin(fourways_neswx), std::end(fourways_neswx), std::begin({0,0,1,1,0,0,1,0})))
+                plaza_dir = 4;
+            else if (std::equal(std::begin(fourways_neswx), std::end(fourways_neswx), std::begin({1,1,1,0,1,1,0,0})))
+                plaza_dir = 3;
+            else if (std::equal(std::begin(fourways_neswx), std::end(fourways_neswx), std::begin({1,1,0,1,1,0,0,1})))
+                plaza_dir = 2;
+            else if (std::equal(std::begin(fourways_neswx), std::end(fourways_neswx), std::begin({1,0,1,1,0,0,1,1})))
+                plaza_dir = 1;
+            else if (std::equal(std::begin(fourways_neswx), std::end(fourways_neswx), std::begin({0,1,1,1,0,1,1,0})))
```

Comment:
Same thing with helpers.
You could easily extract a helper here, maybe even an array of pairs `bool[], dir`.